include/vapor/ArbitrarilyOrientedRegularGrid.h

```diff
@@ -0,0 +1,103 @@
+#pragma once
+
+#include <ostream>
+#include <vector>
+#include <vapor/common.h>
+#include <vapor/RegularGrid.h>
+#include <glm/glm.hpp>
+
+// clang-format off
+
+struct planeDescription {
+    size_t sideSize;
+    std::vector<double> origin;
+    std::vector<double> rotation;
+    VAPoR::CoordType boxMin;
+    VAPoR::CoordType boxMax;
+    VAPoR::CoordType domainMin;
+    VAPoR::CoordType domainMax;
+};
+
+namespace VAPoR {
+
+//! Create a 2D RegularGrid that is rotated along an arbitrary orientation.  This new grid
+//! samples a given 3D Grid object along a plane, through the entire domain.  The "planeDescription" 
+//! variable includes parameters that define the 2D plane's origin, rotation, and valid extents.  Points on the
+//! 2D RegularGrid that are outside of the valid extents will be assigned as
+//! missingValues.
+//!
+//! Note: The boxMin and boxMax values within the planeDescription should come from
+//! the current RenderParams' Box class.  The domainMin and domainMax values should
+//! come from the DataMgr::GetVariableExtents function.  Taking these values from
+//! the given 3D grid can produce erroneous results when the 3D grid's region is reduced.
+//!
+//! \param[in] grid3d A variable's 3D grid to be sampled along a plane, to generate a 2D grid object
+//! \param[in] description A set of std::vector<dobule> values that define the origin, rotation, and extents of the returned 2D grid
+//! \param[in] dims The sampling rate that the new grid will be defined upon, along its X and Y axes
+//! \param[out] data An array of sideSize*sideSize values that contain floating point values of the sampled data.
+//!
+
+class VDF_API ArbitrarilyOrientedRegularGrid : public RegularGrid {
+public:
+    ArbitrarilyOrientedRegularGrid(
+        const VAPoR::Grid* grid3d,
+        planeDescription& pd,
+        const DimsType& dims,
+        std::shared_ptr<float>& blks
+    );
+
+    ArbitrarilyOrientedRegularGrid() = default;
+    virtual ~ArbitrarilyOrientedRegularGrid() = default;
+
+    static std::string GetClassType() { return ("ArbitrarilyOrientedRegular"); }
+    std::string        GetType() const override { return (GetClassType()); }
+
+    //! \copydoc Grid::GetUserCoordinates()
+    //
+    virtual void GetUserCoordinates(const DimsType &indices, CoordType &coords) const override;
+
+    //! Retrieve the winding order that pertains to the generated array of data, if it is
+    //! to be drawn as a texture as is done in the SliceRenderer's vertex buffer.  The winding order is a series
+    //! of six 3D vertices that describe two triangles that define how the texture is drawn in space.
+    void GetWindingOrder( std::vector<double>& windingOrder ) const;
+
+    //! Retrieve the 3D coordinates of the corners of the sampled data.  This is a set of four 3D coordinates,
```

Comment:
Isn't this what Grid::GetUserExtents() should return? Is there a need to introduce a new function, Get3DRectangle()

Reply:
No.  Get3DRectangle() returns four vertices at the corners of the sampled plane needed for the texture coordinate buffer.  

If we have a plane that is rotated 45 degrees on X, GetUserExtents() would return two 3D coordinates that describe the volume enclosing the plane, which isn't what's needed.  The 3D rectangle also cannot be inferred from this volume because a rotation of -45 degrees would return the same volume but describe a different plane.

Reply:
PS - [See here](https://github.com/NCAR/VAPOR/blob/dd7ac4ee9d3aee8e508b6a40723e83c548ef9337/lib/render/SliceRenderer.cpp#L101) for where this structure is needed in the SliceRenderer.

Reply:
OK. So Get3DRectangle() returns the coordinates of the four corner vertices. So this is a convenience function that is equivalent to calling Grid::GetUserCoordinates() with the indices of the four corner points?

Reply:
Almost.  This is hard to explain without a picture.  Do you have time for a screen share?  If not I can try to explain here.

Reply:
The ordering of the four corners of the texture has been moved from the VDC library to the SliceRenderer, at SliceRenderer.cpp:223.  I think this still may be a problematic solution, and requires more discussion.  However it does move renderer logic outside of the VDC library as we discussed.